central/policy/datastore/singleton.go

```diff
@@ -44,20 +46,65 @@ func Singleton() DataStore {
 // addDefaults adds the default policies into the postgres table for policies.
 // TODO: ROX-11279: Data migration for postgres should take care of removing default policies in the bolt bucket named removed_default_policies
 // from the policies table in postgres
-func addDefaults(s policyStore.Store, categoriesDS categoriesDS.DataStore) {
+func addDefaults(s policyStore.Store, categoriesDS categoriesDS.DataStore, fullStore DataStore) {
 	policyIDSet := set.NewStringSet()
+	storedPolicies := make([]*storage.Policy, 0)
 	err := s.Walk(workflowAdministrationCtx, func(p *storage.Policy) error {
 		policyIDSet.Add(p.GetId())
 		// Unrelated to adding/checking default policies, this was put here to prevent looping through all policies a second time
 		if p.GetSource() == storage.PolicySource_DECLARATIVE {
 			metrics.IncrementTotalExternalPoliciesGauge()
 		}
+		storedPolicies = append(storedPolicies, p)
 		return nil
 	})
+
 	if err != nil {
 		panic(err)
 	}
 
+	// ROX-31406: Fix categories that were impacted by previous bug
+	for _, p := range storedPolicies {
+		var categories []*storage.PolicyCategory
+		categories, err = categoriesDS.GetPolicyCategoriesForPolicy(workflowAdministrationCtx, p.GetId())
```

Comment:
We could save this query per policy if we do this in the walk. Create a list of shouldReUpsert policies right there in the walk? and upsert them after.